sunpy/map/mapbase.py

```diff
@@ -1464,23 +1471,89 @@ def submap(self, bottom_left, top_right=None):
+        Exposure Time:		 0.234256 s
+        Dimension:		 [4. 5.] pix
+        Coordinate System:	 helioprojective
+        Scale:			 [2.402792 2.402792] arcsec / pix
+        Reference Pixel:	 [1.5 1.5] pix
+        Reference Coord:	 [3.22309951 1.38578135] arcsec
+        array([[213.9748 , 256.76974, 244.54262, 356.62466],
+            [223.74321, 258.0102 , 292.27716, 340.65408],
+            [219.53459, 242.31648, 308.5911 , 331.373  ],
+            [268.24377, 254.83157, 268.24377, 321.89252],
+            [249.99167, 265.14267, 274.61206, 240.5223 ]], dtype=float32)
+        """
+        if not isinstance(bottom_left, u.Quantity):
+            bottom_left, top_right = get_rectangle_coordinates(bottom_left,
+                                                               top_right=top_right,
+                                                               width=width,
+                                                               height=height)
 
             bottom_left = self.world_to_pixel(bottom_left)
             top_right = self.world_to_pixel(top_right)
 
-        elif not (isinstance(bottom_left, u.Quantity) and bottom_left.unit.is_equivalent(u.pix) and
-                  isinstance(top_right, u.Quantity) and top_right.unit.is_equivalent(u.pix)):
-            raise ValueError("Invalid input, bottom_left and top_right must either be SkyCoord or Quantity in pixels.")
+        elif (
+                # You haven't given top_right or both width and height
+                (top_right is None and (width is None or height is None)) or
+                # You have given top_right but you have also given width and height
+                (top_right is not None and (width is not None or height is not None)) or
+                # You haven't given top_right, width or height
+                (top_right is None and width is None and height is None)
+        ):
+            raise ValueError("If bottom_left is not a SkyCoord either top_right or "
+                             "both width and height must be specified.")
+        elif (
+                # bottom_left is not a Quantity in pixels
+                (not isinstance(bottom_left, u.Quantity) or not bottom_left.unit.is_equivalent(u.pix)) or
+                # You gave top_right but it's not a Quantity in pixels
+                (top_right is not None and (not isinstance(top_right, u.Quantity) or not top_right.unit.is_equivalent(u.pix))) or
+                # You didn't give top_right and width and height aren't both quantities in pixels
+                (top_right is None and ((not isinstance(width, u.Quantity) or not width.unit.is_equivalent(u.pix)) and
+                                        (not isinstance(height, u.Quantity) or not height.unit.is_equivalent(u.pix))))
+        ):
+            raise TypeError("When bottom_left is not a SkyCoord any values of top_right, "
+                             "width or height specified must be Quantity objects in units of pixels.")
```

Comment:
```suggestion
        elif (not all([arg is None or (isinstance(arg, u.Quantity) and arg.unit.is_equivalent(u.pix))
                       for arg in (bottom_left, top_right, width, height)]))):
            raise TypeError("When bottom_left is not a SkyCoord, any values of top_right, "
                             "width or height specified must be Quantity objects in units of pixels.")
```
The correct `None` pattern has already been checked, so this check doesn't need all of the logic branches.